alf/metrics/metrics.py

```diff
@@ -408,6 +416,118 @@ def _extract_and_process_acc_value(self, acc, last_episode_indices):
             last_episode_indices]
 
 
+@alf.configurable
+class EpisodicStartAverageDiscountedReturnMetric(
+        AverageDiscountedReturnMetric):
+    """Metric for computing the discounted return from episode start states.
+    It is calculated according to the following formula:
+
+    .. math::
+        \begin{array}{ll}
+            R &=r_1 + \gamma r_2 + \gamma^2 r_3 + \cdots \\
+            &= \sum_{l=1}^L \gamma^{l-1} r_l,
+        \end{array}
+    where :math:`\gamma` is the reward discount, and :math:`r_1` denotes the
+    reward due to the first action, which is received at the second time step.
+    :math:`L` equals to the episode length - 1.
+
+
+    Note that if the last step is not due to time limit, the discounted return
+    calculated from the formula above is unbiased. If the last step is due to
+    time limit, it is a biased estimate and its expectation is lower than the
+    ground-truth (when rewards are non-negative).
+    """
+
+    def __init__(self,
+                 example_time_step: TimeStep,
+                 name='EpisodicStartAverageDiscountedReturn',
+                 prefix='Metrics',
+                 buffer_size=10,
+                 reward_clip=None):
+        super(EpisodicStartAverageDiscountedReturnMetric, self).__init__(
+            name=name,
+            prefix=prefix,
+            buffer_size=buffer_size,
+            example_time_step=example_time_step,
+            reward_clip=reward_clip)
+
+    def _extract_metric_values(self, time_step):
+        """Accumulate discounted immediate rewards to get discounted episodic
+        return. It also updates the accumulated discount and step count.
+        """
+        self._update_discount_and_step_count(time_step)
+
+        ndim = time_step.step_type.ndim
+        if time_step.reward.ndim == ndim:
+            discounted_reward = time_step.reward * self._accumulated_discount
+        else:
+            reward = time_step.reward.reshape(*time_step.step_type.shape, -1)
+            discounted_reward = [
+                reward[..., i] * self._accumulated_discount
+                for i in range(reward.shape[-1])
+            ]
```

Comment:
`discounted_reward = list(reward * self._accumulated_discount)`

Reply:
Done.  It has to be permuted:
`            discounted_reward = list(
                (reward * self._accumulated_discount).permute(
                    reward.ndim - 1, *torch.arange(reward.ndim - 1)
                ))
`